WorkspaceServer/Servers/Roslyn/TrackingStringWriter.cs

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace WorkspaceServer.Servers.Roslyn
+{
+    internal class TrackingStringWriter : StringWriter
+    {
+        private class Region
+        {
+            public int Start { get; set; }
+            public int Length { get; set; }
+        }
+         List<Region> _regions = new List<Region>();
+         private bool _trakingSize;
+
+
+         public bool WriteOccurred { get; set; }
+
+        public override void Write(char value)
+        {
+            TrackSize(() => base.Write(value));
```

Comment:
You're going to reallocate this delegate every time you write, consider caching it.